lib/fog/aws/requests/compute/create_network_interface.rb

```diff
@@ -82,12 +83,14 @@ def create_network_interface(subnetId, options = {})
               end
 
               if options['PrivateIpAddress'].nil?
+                range = cidr_block.to_range
                 # Here we try to act like a DHCP server and pick the first
                 # available IP (not including the first in the cidr block,
                 # which is typically reserved for the gateway).
-                cidr_block.each_host do |p_ip|
-                  unless self.data[:network_interfaces].map{ |ni, ni_conf| ni_conf['privateIpAddress'] }.include?p_ip.to_s ||
-                    cidr_block.first == p_ip
+                range = range.drop(2)[0..-2] if cidr_block.ipv4?
```

Comment:
Sorry, I'm not super familiar with all of this. Could you help me understand what the drop and slice are needed for here? I tried to look at the IPAddress code for each_host:

https://github.com/ipaddress-gem/ipaddress/blob/8ff186bd8d71da69fa33b999ff69be9d8885c424/lib/ipaddress/ipv4.rb#L448-L452

I can see the +1 and -1 there, which I imagine omit the first and last things in the range, but I think drop(2) and -2 would actually omit 2 from the front and back instead? Certainly possibly that I'm missing something here, but wanted to try and ensure I understood what we expected.

Reply:
It was also a bit vague to me, but it has to do with the other comment you placed. In my examples I will use `192.0.2.0/24` as the subnet.

You're right about +1 to drop the network address (`192.0.2.0`). Then the additional +1 is for the router (see https://github.com/fog/fog-aws/pull/630/files#r803550336). That makes +2 and is `.drop(2)`.

The -1 is the broadcast address (`192.0.2.255`). The reason I use `-2` in there is just Ruby. `-1` is the end, so `[0..-1]` is the same as not putting anything in.
```ruby
x = ['a', 'b', 'c']
assert x[0..-1] == x
```

But perhaps it's better to essentially implement `to_range` manually here and use the integer representation of addresses.

Reply:
I took a stab at this and something like this will work with ipaddr 1.2.3:
```ruby
def usable_addresses(block)
  if block.ipv4?
    mask_addr = IPAddr.new(block.netmask)
    # Exclude the network address and the first IP,
    # which is typically reserved for the gateway).
    begin_addr = (block & mask_addr).to_i + 2
    # Exclude the broadcast address
    end_addr = (block.to_i | (IPAddr::IN4MASK ^ mask_addr.to_i)) - 1

    IPAddr.new(begin_addr, block.family)..IPAddr.new(end_addr, block.family)
  else
    block.to_range
  end
end
```

Sadly, this needs 1.2.3 which isn't in Ruby 3.0 at least. It's because `netmask` as only added in https://github.com/ruby/ipaddr/commit/283d16f3a379d440328e0eb6b8675950988aa2cb.

You could add it as a dependency. You'll still be trading one for the other. The alternative is to hack around it get a private instance variable.

Reply:
Yeah, thanks for the additional detail. At the end of the day this functionality is in the mocks and probably/hopefully not super impactful, so hopefully we are fine to proceed.